File: apps/web-console/lib/api/audit.ts
/** /v1/audit client. */

export interface AuditLogEntry {
  id: number;
  actor_id: string;
  action: string;
  resource_type: string | null;
  resource_id: string | null;
  detail: Record<string, unknown>;
  ip: string | null;
  user_agent: string | null;
  created_at: string;
}

export interface AuditList {
  items: AuditLogEntry[];
  next_cursor: number | null;
}

export class AuditApi {
  constructor(private apiKey: string) {}
  private async req<T>(path: string, init?: RequestInit): Promise<T> {
    const r = await fetch(path, {
      ...init,
      headers: {
        ...(init?.headers ?? {}),
        Authorization: `Bearer ${this.apiKey}`,
        "Content-Type": "application/json",
      },
    });
    if (!r.ok) throw new Error(`${r.status}: ${(await r.text()).slice(0, 100)}`);
    return r.json() as Promise<T>;
  }
  list(opts: {
    actor_id?: string;
    action_prefix?: string;
    resource_type?: string;
    resource_id?: string;
    since?: string;        // ISO datetime
    until?: string;
    limit?: number;
    before_id?: number;
  } = {}): Promise<AuditList> {
    const p = new URLSearchParams();
    if (opts.actor_id) p.set("actor_id", opts.actor_id);
    if (opts.action_prefix) p.set("action_prefix", opts.action_prefix);
    if (opts.resource_type) p.set("resource_type", opts.resource_type);
    if (opts.resource_id) p.set("resource_id", opts.resource_id);
    if (opts.since) p.set("since", opts.since);
    if (opts.until) p.set("until", opts.until);
    if (opts.limit) p.set("limit", String(opts.limit));
    if (opts.before_id) p.set("before_id", String(opts.before_id));
    return this.req(`/api/audit?${p.toString()}`);
  }

  /** 触发浏览器下载 CSV; 用 fetch + blob 而非直接 a.href, 因为要带 Bearer header */
  async downloadCsv(opts: {
    actor_id?: string;
    action_prefix?: string;
    resource_type?: string;
    resource_id?: string;
    since?: string;
    until?: string;
    max_rows?: number;
  } = {}): Promise<void> {
    const p = new URLSearchParams();
    if (opts.actor_id) p.set("actor_id", opts.actor_id);
    if (opts.action_prefix) p.set("action_prefix", opts.action_prefix);
    if (opts.resource_type) p.set("resource_type", opts.resource_type);
    if (opts.resource_id) p.set("resource_id", opts.resource_id);
    if (opts.since) p.set("since", opts.since);
    if (opts.until) p.set("until", opts.until);
    if (opts.max_rows) p.set("max_rows", String(opts.max_rows));
    const r = await fetch(`/api/audit/csv?${p.toString()}`, {
      headers: { Authorization: `Bearer ${this.apiKey}` },
    });
    if (!r.ok) throw new Error(`${r.status}: ${(await r.text()).slice(0, 100)}`);
    const blob = await r.blob();
    const url = URL.createObjectURL(blob);
    const a = document.createElement("a");
    a.href = url;
    const cd = r.headers.get("content-disposition") || "";
    const m = /filename="([^"]+)"/.exec(cd);
    a.download = m?.[1] ?? "audit.csv";
    document.body.appendChild(a);
    a.click();
    document.body.removeChild(a);
    setTimeout(() => URL.revokeObjectURL(url), 1000);
  }

  cleanup(retainDays: number): Promise<{ deleted: number; retain_days: number }> {
    return this.req(`/api/audit/cleanup`, {
      method: "POST",
      body: JSON.stringify({ retain_days: retainDays }),
    });
  }
}

File: services/api-server/src/api_server/routes/audit.py
"""Audit log read endpoints + CSV 导出 + retention 清理.

GET    /v1/audit             分页 list (含 since/until/actor/action/resource_type)
GET    /v1/audit.csv         同上, 但流式返回 CSV (含 BOM 让 Excel 正确显示中文)
POST   /v1/audit/cleanup     删 retain_days 天前的记录 (返回 deleted 数)
"""

from __future__ import annotations

import asyncio
import csv
import io
import json
from datetime import datetime
from typing import Any
from uuid import UUID

from fastapi import APIRouter, Depends, Query, Request
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, Field

from api_server.acl import Principal
from api_server.auth import verify_api_key
from api_server.db import audit as audit_db

router = APIRouter()


class AuditOut(BaseModel):
    id: int
    actor_id: str
    action: str
    resource_type: str | None
    resource_id: str | None
    detail: dict[str, Any]
    ip: str | None
    user_agent: str | None
    created_at: datetime


class AuditList(BaseModel):
    items: list[AuditOut]
    next_cursor: int | None = None


def _to_out(r: audit_db.AuditRow) -> AuditOut:
    return AuditOut(
        id=r.id, actor_id=r.actor_id,
        action=r.action,
        resource_type=r.resource_type,
        resource_id=r.resource_id,
        detail=r.detail, ip=r.ip,
        user_agent=r.user_agent,
        created_at=r.created_at,
    )


@router.get("/audit", response_model=AuditList)
async def list_audit(
    actor_id: str | None = Query(None),
    action_prefix: str | None = Query(None),
    resource_type: str | None = Query(None),
    resource_id: str | None = Query(None),
    since: datetime | None = Query(None, description="ISO 时间, >= 该时刻"),
    until: datetime | None = Query(None, description="ISO 时间, < 该时刻"),
    limit: int = Query(50, ge=1, le=200),
    before_id: int | None = Query(None),
    principal: Principal = Depends(verify_api_key),
) -> AuditList:
    rows, next_cursor = await asyncio.to_thread(
        audit_db.list_recent,
        tenant_id=principal.tenant_id,
        actor_id=actor_id,
        action_prefix=action_prefix,
        resource_type=resource_type,
        resource_id=resource_id,
        since=since,
        until=until,
        limit=limit,
        before_id=before_id,
    )
    return AuditList(items=[_to_out(r) for r in rows], next_cursor=next_cursor)


@router.get("/audit/csv")
async def export_audit_csv(
    actor_id: str | None = Query(None),
    action_prefix: str | None = Query(None),
    resource_type: str | None = Query(None),
    resource_id: str | None = Query(None),
    since: datetime | None = Query(None),
    until: datetime | None = Query(None),
    max_rows: int = Query(5000, ge=1, le=50000),
    principal: Principal = Depends(verify_api_key),
):
    """导出审计为 CSV. 默认最多 5000 行 (max_rows ≤ 50000).

    含 UTF-8 BOM, Excel 打开不乱码.
    """
    rows, _ = await asyncio.to_thread(
        audit_db.list_recent,
        tenant_id=principal.tenant_id,
        actor_id=actor_id,
        action_prefix=action_prefix,
        resource_type=resource_type,
        resource_id=resource_id,
        since=since,
        until=until,
        limit=max_rows,
        before_id=None,
    )

    def gen():
        buf = io.StringIO()
        # UTF-8 BOM 让 Excel 正确识别中文 CSV
        buf.write("\ufeff")
        w = csv.writer(buf, quoting=csv.QUOTE_MINIMAL)
        w.writerow([
            "id", "created_at", "actor_id", "action",
            "resource_type", "resource_id", "ip", "user_agent",
            "detail_json",
        ])
        yield buf.getvalue()
        buf.seek(0); buf.truncate(0)
        for r in rows:
            w.writerow([
                r.id,
                r.created_at.isoformat(),
                r.actor_id,
                r.action,
                r.resource_type or "",
                r.resource_id or "",
                r.ip or "",
                (r.user_agent or "").replace("\n", " "),
                json.dumps(r.detail, ensure_ascii=False, separators=(",", ":")),
            ])
            yield buf.getvalue()
            buf.seek(0); buf.truncate(0)

    fname = f"audit-{datetime.utcnow().strftime('%Y%m%d-%H%M%S')}.csv"
    return StreamingResponse(
        gen(),
        media_type="text/csv; charset=utf-8",
        headers={"Content-Disposition": f'attachment; filename="{fname}"'},
    )


class CleanupRequest(BaseModel):
    retain_days: int = Field(..., ge=1, le=3650, description="保留多少天")


class CleanupResponse(BaseModel):
    deleted: int
    retain_days: int


@router.post("/audit/cleanup", response_model=CleanupResponse)
async def cleanup_audit(
    payload: CleanupRequest,
    request: Request,
    principal: Principal = Depends(verify_api_key),
) -> CleanupResponse:
    """删 retain_days 天前的 audit. 自身行为也会被记录 (meta-audit)."""
    n = await asyncio.to_thread(
        audit_db.cleanup_older_than,
        tenant_id=principal.tenant_id,
        retain_days=payload.retain_days,
    )
    # meta-audit
    ip = (
        request.headers.get("x-forwarded-for", "").split(",")[0].strip()
        or (request.client.host if request.client else None)
    )
    audit_db.insert(
        tenant_id=principal.tenant_id, actor_id=principal.actor_id,
        action="audit.cleanup",
        resource_type="audit_log", resource_id=None,
        detail={"retain_days": payload.retain_days, "deleted": n},
        ip=ip, user_agent=request.headers.get("user-agent"),
    )
    return CleanupResponse(deleted=n, retain_days=payload.retain_days)
